services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/utils/MongoDbDocumentBuilder.java

```diff
@@ -25,6 +27,10 @@
  */
 public final class MongoDbDocumentBuilder {
```

Comment:
I don't understand the concept of this class. Why not simply create static methods, returning `JsonObject`, looking at the rest of the locations, this is all what is required.

Reply:
The methods in this class to initialize the JsonObject (e.g.: `forTenantId` )query are already static. 
I think the main purpose of the class is to simplify the handling of jsonObject and property key constants.

Reply:
To me this makes it look more complicated. You have a static method, which creates and instance of an object, which fills a JSON object. Later on you ditch the object, and only use the JSON object.

Reply:
When just a JsonObject is returned, we lose the abillty to make easy to read query creation chains:
```
final JsonObject updateTenantQuery = MongoDbDocumentBuilder
    .forVersion(resourceVersion)
    .withTenantId(tenantId)
    .document();
```
would change to e.g.:
```
JsonObject updateTenantQuery = new JsonObject();
if (resourceVersion.isPresent()) {
    updateTenantQuery = MongoDbDocumentBuilder.withVersion(updateTenantQuery, tenantId);
}
updateTenantQuery = MongoDbDocumentBuilder.withTenantId(updateTenantQuery, resourceVersion);
```

Reply:
Yes, but you have `withTenantId` working on the instance, and a static `forTenantId` which only wraps around `withTenantId`. I would propose to remove one level of wrapping:

```java
final JsonObject updateTenantQuery = MongoDbDocumentBuilder
    .builder()
    .withVersion(resourceVersion)
    .withTenantId(tenantId)
    .document();
```

That would allow you drop all (duplicate) static wrapper methods, and still use chained method invocations.

Or inline the `with` methods, and focus on the static `for` methods. Having both doesn't make sense to me.

Reply:
To me it looks reasonable to have both the factory methods for creating the instance based on a resource version or tenant ID as well as the methods for adding additional properties. That way you only need `.forTenantId("tenant").withVersion("1")` instead of `.builder().withTenantId("tenant").withVersion("1")`.
I don't see a problem here.

Reply:
I see the arguments from @ctron, but I find the the old implementation in the services are more readable.

IMHO the problem is that an optional version can only be added conditionally `withVersion()`, which expands the code.
```
deleteDeviceQuery =
  resourceVersion.map(v -> MongoDbDocumentBuilder.builder().withVersion(v))
  .orElse(MongoDbDocumentBuilder.builder())
    .withTenantId(tenantId)
    .withDeviceId(deviceId)
    .document();
```

My suggestion would be to remove the factory `for*` methods, as @ctron has suggested to simplify `MongoDbDocumentBuilder`.
In case of an optional version, I would introduce a new method `withOptionalVersion()`.